### PR TITLE
Integrating enumerables methods 

### PR DESCRIPTION
**Pull request by luisvinicius09**
### This code includes 9 methods.
* `each` method.
* `each_with_index` method.
* `select` method.
* `all?` method.
* `any?` method.
* `none?` method.
* `count` method.
* `map` method.
* `inject` method.

###### - `multiply_els` method.

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      end
+    end
+    count
+  end
+end
+
+# This method is only for tests
+def multiply_els(args)
+  args.my_inject { |acc, val| acc * val }
+end
```
Comment (kevcap) [inline enumerables.rb:161]: In this scenario, we got this error:
rspec ./spec/enumerables_spec.rb:460 # enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      my_each do |x|
+        y = y ? y.send(s, x) : val
+      end
+    else
+      my_each do |x|
+        y = y ? yield(y, x) : val
+      end
+    end
+    y
+  end
```
Comment (kevcap) [inline enumerables.rb:137]: At the my inject one, we got a few errors, please take a look and fix it:

1. rspec ./spec/enumerables_spec.rb:405 # enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
2. rspec ./spec/enumerables_spec.rb:409 # enumerables #my_inject searches for the longest word in an array of strings
3. rspec ./spec/enumerables_spec.rb:413 # enumerables #my_inject does not mutate the original array 
4. rspec ./spec/enumerables_spec.rb:419 # enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block   
5. rspec ./spec/enumerables_spec.rb:423 # enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
6. rspec ./spec/enumerables_spec.rb:431 # enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block      
7. rspec ./spec/enumerables_spec.rb:439 # enumerables #my_inject when a symbol is specified without 
8. an initial value combines each element of the collection by applying the symbol as a named methodrspec ./spec/enumerables_spec.rb:443 # enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
9. rspec ./spec/enumerables_spec.rb:449 # enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
10. rspec ./spec/enumerables_spec.rb:453 # enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      my_each do |x|
+        new_array.push(x) if yield(x)
+      end
+    end
+    new_array
+  end
+
+  def my_all?(args = nil)
+    return to_enum unless block_given? || !args.nil?
+
```
Comment (kevcap) [inline enumerables.rb]: At the my_all?, we got these errors:

1. - rspec ./spec/enumerables_spec.rb:161 # enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
2. - rspec ./spec/enumerables_spec.rb:165 # enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
3. - rspec ./spec/enumerables_spec.rb:202 # enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      my_each { |x| return true if x.is_a?(args) }
+    else
+      my_each { |x| return true if x == args }
+    end
+    false
+  end
+
+  def my_none?(args = nil)
+    if args.nil?
+      if block_given?
```
Comment (kevcap) [inline enumerables.rb:39]: At the my_none?, we got these errors:

1. rspec ./spec/enumerables_spec.rb:330 # enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
2. rspec ./spec/enumerables_spec.rb:333 # enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    elsif args.is_a?(Module)
+      my_each { |x| return false if x.is_a?(args) }
+    else
+      my_each { return false if val == args }
+    end
+    true
+  end
+
+  def my_select
+    return to_emum :my_select unless block_given?
```
Comment (kevcap) [inline enumerables.rb]: At my_select we only got one error:
1. rspec ./spec/enumerables_spec.rb:128 # enumerables #my_select returns an enumerator if no block is given

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
## [REQUIRED CHANGES]
- [ ] We just have a few errors, which I typed down the reasons of the errors;

Tip: Remember to use a lot of types of inputs such as arrays, strings, words (w%[ ... ]), numbers, hashes, ranges (Range.new), blocks and even the array.clone method.

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In the real-world, you'll have a lot of change requests for your real clients, it's good to get used to it.</sub>

File: enumerables.rb (hunk 5 above)
Comment (luisvinicius09) [reply enumerables.rb]: Sorry, I don't understand why that is an error if the [docs](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-select) is saying:
`If no block is given, an Enumerator is returned instead.`
<br>
#### I appreciate it if you explain to me.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @luisvinicius09 and @GuNner-pete, Great job! 👏  Both of you have done well on this project.

However, there are still some changes that are necessary for this project to be approved.

- [x] `#my_select` returns an enumerator if no block is given
`[1,2,3,4].my_select.class==Enumerator`

- [x] `#my_all?` when no block or argument is given returns true when none of the collection members are false or nil
 `[1, true, "hi", []].all? ==  [1, true, "hi", []].my_all?`

- [x] `#my_all?` when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
`[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`

- [x] `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`['dog','car'].none?(5) == ['dog','car'].my_none?(5)`

- [x] `#my_none?` when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
`%w[dog door rod blade].none?(5) == %w[dog door rod blade].my_none?(5)`

- [ ] `#my_inject`` raises a "LocalJumpError" when no block or argument is given
     `Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
       expected LocalJumpError but nothing was raised`

- [x] `#my_inject` searches for the longest word in an array of strings
`['to','four','tri'].my_inject(&proc{|a,b| [a,b].max_by{|m| m.length}}) == ['to','four','tri'].inject(&proc{|a,b| [a,b].max_by{|m| m.length}})`

- [x] `#my_inject` does not mutate the original array 
```
array = [1,2,3,4]
array_clone = array.clone
array.my_inject{|total, num| total+num}
array == array_clone
```

- [x]   `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
`array.my_inject(&operation).to eq array.inject(&operation)`

- [x] `#my_inject` when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(:+) == [1,2,3].my_inject(:+)`

 
- [x] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
`(1..9).inject(:+) == (1..9).my_inject(:+)`

- [x] `#my_inject` when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(4, :*) == [1,2,3].my_inject(4, :*)`

- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
`(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`

 
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
`multiply_els([1,2,3]) == 6`

### Kind make these changes and request for another code review

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

A great effort with the changes, team! 👏 
But we still need to implement one more change to get the project approved 💪 
Check my comments below, apply the change and resubmit your project for a new code review ⏩ 
Good luck! 😺 
# Mandatory changes
- #my_inject should raise a "LocalJumpError" when no block or argument is given

**Review by ghost — APPROVED** [review]
# Status: Approved 🎉 

**Congratulations team and good luck with your next project 😄.**

**Pull request merged**